File: subsystems/gripper.py
"""Gripper subsystem -- spinning roller intake at the end of the elevator.

Teaching-bot proof of concept. Same intake/eject pattern as the competition
bot's Feeder subsystem, just one motor instead of two.
"""
from __future__ import annotations

import rev
from commands2 import Command, Subsystem, cmd

from constants import GripperConstants


class Gripper(Subsystem):
    def __init__(self) -> None:
        super().__init__()

        self._roller_motor = rev.SparkMax(GripperConstants.ROLLER_MOTOR_ID, rev.SparkMax.MotorType.kBrushless)
        roller_config = rev.SparkMaxConfig()
        roller_config.inverted(GripperConstants.ROLLER_MOTOR_INVERTED)
        roller_config.setIdleMode(rev.SparkBaseConfig.IdleMode.kBrake)
        roller_config.smartCurrentLimit(GripperConstants.ROLLER_CURRENT_LIMIT)
        self._roller_motor.configure(
            roller_config, rev.ResetMode.kResetSafeParameters, rev.PersistMode.kPersistParameters
        )

    def stop(self) -> None:
        self._roller_motor.set(0.0)

    def intake_command(self) -> Command:
        return cmd.run(lambda: self._roller_motor.set(GripperConstants.INTAKE_SPEED), self).finallyDo(
            lambda interrupted: self.stop()
        )

    def eject_command(self) -> Command:
        return cmd.run(lambda: self._roller_motor.set(GripperConstants.EJECT_SPEED), self).finallyDo(
            lambda interrupted: self.stop()
        )

File: src/main/java/frc/robot/subsystems/Gripper.java
package frc.robot.subsystems;

import static frc.robot.Constants.GripperConstants.*;

import com.revrobotics.PersistMode;
import com.revrobotics.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.config.SparkMaxConfig;

import edu.wpi.first.wpilibj2.command.SubsystemBase;

/**
 * Gripper subsystem -- spinning roller intake at the end of the elevator.
 *
 * <p>Teaching-bot proof of concept. The simplest subsystem here: one motor, no sensors
 * at all. {@code setSpeed()}/{@code stop()} are the only two things it knows how to do
 * -- commands/IntakeCommand.java and commands/EjectCommand.java just pick which speed
 * to hold while a button is pressed.
 */
public class Gripper extends SubsystemBase {

    private final SparkMax rollerMotor = new SparkMax(ROLLER_MOTOR_ID, MotorType.kBrushless);

    /**
     * No parameters besides the implicit {@code this} here -- Gripper doesn't need
     * anything handed to it from the outside to build itself; every value it needs
     * (motor CAN ID, current limit, ...) comes from {@code GripperConstants} instead.
     * Compare this to commands/IntakeCommand.java's constructor, which DOES take a
     * parameter ({@code Gripper gripper}) because a command needs to be told WHICH
     * Gripper object to act on.
     */
    public Gripper() {
        SparkMaxConfig rollerConfig = new SparkMaxConfig();
        rollerConfig.inverted(ROLLER_MOTOR_INVERTED);
        rollerConfig.idleMode(IdleMode.kBrake);
        rollerConfig.smartCurrentLimit(ROLLER_CURRENT_LIMIT);
        rollerMotor.configure(rollerConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    }

    /**
     * {@code speed} is a duty cycle in [-1, 1]: positive intakes, negative ejects -- see
     * {@code GripperConstants.INTAKE_SPEED}/{@code EJECT_SPEED}. The parameter is
     * written {@code double speed} rather than {@code speed: float} the way Python
     * wrote it -- Java always puts the type BEFORE the name, with no colon, for every
     * parameter and every field in this project; Python puts the type AFTER the name,
     * with a colon, and only when someone chooses to add the (optional) hint.
     */
    public void setSpeed(double speed) {
        rollerMotor.set(speed);
    }

    public void stop() {
        rollerMotor.set(0.0);
    }
}
